### PR TITLE
typings: Refactor how channel types are done

### PR DESCRIPTION
**Pull request by didinele**
**Please describe the changes this PR makes and why it should be merged:**
Inspired by #3807, originally going for a different solution but ultimately becoming a copy due to certain issues I ran into

Changes made:
- Each Channel class now has a corresponding type property
- GuildCreateChannelOptions#type now actually takes numbers

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [x] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: typings/index.d.ts [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-	store,
-	unknown
+	TEXT = 'text',
+	DM = 'dm',
+	VOICE = 'voice',
+	GROUP = 'group',
+	CATEGORY = 'category',
+	NEWS = 'news',
+	STORE = 'store',
+	UNKNOWN = 'unknown'
```
Comment (pizzafox) [inline typings/index.d.ts]: TypeScript standard it to use pascal-case for enums ([source: TypeScript docs](typescriptlang.org/docs/handbook/enums.html#string-enums)). Enums are inherently constant, so it doesn't really matter that we use uppercase here.

**Comment by didinele** [comment]
Oh yeah, sorry about that, if you look at the commit log I had to force push at some point as the branch was messed up, I was doing it full uppercase initially, will do it when I get home tomorrow.

File: typings/index.d.ts [hunk 2]
```diff
@@ -2291,7 +2297,7 @@ declare module 'discord.js' {
 	interface GuildCreateChannelOptions {
 		permissionOverwrites?: OverwriteResolvable[] | Collection<Snowflake, OverwriteResolvable>;
 		topic?: string;
-		type?: Exclude<keyof typeof ChannelType, 'dm' | 'group' | 'unknown'>;
+		type?: Exclude<ChannelType, ChannelType.DM | ChannelType.GROUP | ChannelType.UNKNOWN>;
```
Comment (vladfrangu) [inline typings/index.d.ts]: Won't this change prevent using numbers for the type? Did my old change cause that too?

Comment (didinele) [reply typings/index.d.ts]: That's a good point Vlad, it will.

Not sure what to do about this, honestly.

Comment (vladfrangu) [reply typings/index.d.ts]: I mean.. won't reverting the enum change itself fix it? By default, enum entries start at 0, which is also why I made the keys lowercased (to match what djs can parse)

Comment (didinele) [reply typings/index.d.ts]: But then that defeats the whole purpose of this PR, of not being forced to use string literals for the individual `type` properties

Comment (appellation) [reply typings/index.d.ts]: You would still be forced to use either string literals or numbers for type guarding the channels.

Comment (didinele) [reply typings/index.d.ts]: Type guarding yeah, I'm talking about the literal typings

File: typings/index.d.ts (hunk 1 above)
Comment (didinele) [reply typings/index.d.ts]: Your link seems to be invalid, could you toss me the proper one? I cannot seem to find this standard/convention you are talking about.

Comment (pizzafox) [reply typings/index.d.ts]: Sorry, here's the correct link:

https://www.typescriptlang.org/docs/handbook/enums.html#string-enums

You can see in their example codeblocks from that page that they are using Pascal case.

Comment (didinele) [reply typings/index.d.ts]: Despite it commonly being in official documentation, it is never stated that pascal case is the standard.

In fact, there's a few cases on that very page that use full upppercase.

I don't feel like I'm breaking a convention here so unless anyone else has something to say about this I'm leaving it this way.

File: typings/index.d.ts [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
-	STORE = 'store',
-	UNKNOWN = 'unknown'
+	text = 0,
+	dm = 1,
+	voice = 2,
+	group = 3,
+	category = 4,
+	news = 5,
+	store = 6,
+	unknown = 7
```
Comment (pizzafox) [inline typings/index.d.ts]: You don't have to write the numbers here, the default behavior is to increment by one starting at 0.

Comment (didinele) [reply typings/index.d.ts]: Just noticed too, on it

**Pull request merged**